pkg/scheduler/operations.go

```diff
@@ -57,36 +44,198 @@ func IsOperationNotFoundError(err error) bool {
 	return ok
 }
 
-type DefaultOperationsRegistry struct {
-	registry map[string]map[string]OperationState
+type PersistedOperationsRegistry struct {
+	*repository.Repository
+}
+
+func NewPersistedOperationsRegistry(dbFac db.ConnectionFactory, debug bool) (OperationsRegistry, error) {
+	repo, err := repository.NewRepository(dbFac, debug)
+	if err != nil {
+		return nil, err
+	}
+	return &PersistedOperationsRegistry{repo}, nil
+}
+
+func (or *PersistedOperationsRegistry) GetDoneOperations(schedulingID string) ([]*model.OperationEntity, error) {
+	return nil, nil
+}
+
+func (or *PersistedOperationsRegistry) RegisterOperation(correlationID, schedulingID, component string, version int64) (*model.OperationEntity, error) {
+	dbOps := func() (interface{}, error) {
+		opEntity := &model.OperationEntity{
+			SchedulingID:  schedulingID,
+			CorrelationID: correlationID,
+			ConfigVersion: version,
+			Component:     component,
+			State:         model.OperationStateNew,
+		}
+		_, err := or.GetOperation(correlationID, schedulingID)
+		if err == nil {
+			return nil, fmt.Errorf("operation with the following id %s already registered", correlationID)
+		} else if !repository.IsNotFoundError(err) {
+			//unexpected error
+			return nil, err
+		} else {
+			q, err := db.NewQuery(or.Conn, opEntity)
+			if err != nil {
+				return nil, err
+			}
+			err = q.Insert().Exec()
+			if err != nil {
+				return nil, err
+			}
+			return opEntity, nil
+		}
+	}
+	entity, err := db.TransactionResult(or.Conn, dbOps, or.Logger)
+	if err != nil {
+		return nil, err
+	}
+	return entity.(*model.OperationEntity), nil
+}
+
+func (or *PersistedOperationsRegistry) GetOperation(correlationID, schedulingID string) (*model.OperationEntity, error) {
+	q, err := db.NewQuery(or.Conn, &model.OperationEntity{})
+	if err != nil {
+		return nil, err
+	}
+	whereCond := map[string]interface{}{
+		"CorrelationID": correlationID,
+		"SchedulingID":  schedulingID,
+	}
+	opEntity, err := q.Select().
+		Where(whereCond).
+		GetOne()
+	if err != nil {
+		return nil, or.NewNotFoundError(err, opEntity, whereCond)
+	}
+	return opEntity.(*model.OperationEntity), nil
+}
+
+func (or *PersistedOperationsRegistry) RemoveOperation(correlationID, schedulingID string) error {
+	dbOps := func() (interface{}, error) {
+		_, err := or.GetOperation(correlationID, schedulingID)
+		if err != nil {
+			if !repository.IsNotFoundError(err) {
+				return nil, err
+			}
+			return nil, fmt.Errorf("operation with the following id %s not found", correlationID)
+		}
+
+		q, err := db.NewQuery(or.Conn, &model.OperationEntity{})
+		if err != nil {
+			return nil, err
+		}
+		whereCond := map[string]interface{}{
+			"CorrelationID": correlationID,
+			"SchedulingID":  schedulingID,
+		}
+		_, err = q.Delete().
```

Comment:
do you think we should have a kind of ops-archive to have the reconciliation history available for a longe period?

I could images SREs could have a use-case for reviewing the last reconciliation protocl at a later time, what do you think?